sockeye/data_io.py

```diff
@@ -1114,17 +1168,32 @@ def create_sequence_readers(sources: List[str], target: str,
     return source_sequence_readers, target_sequence_reader
 
 
-def parallel_iter(source_iters: Sequence[Iterable[Optional[Any]]], target_iterable: Iterable[Optional[Any]]):
+def parallel_iter(source_iters: Sequence[Iterable[Optional[Any]]],
+                  target_iterable: Iterable[Optional[Any]],
+                  reiterate: bool = True):
```

Comment:
This is a deep change, perhaps. The function needs to be able to continue reading from an already-begun iterator in order to load the data from disk one batch at a time. The test cases pass, but are failing because a list gets returned in one place instead of a tuple. The code is also a bit uglier; I could refactor this into a `paraller_iter_continued()` function or something of that sort. I don't understand why the current code calls `iter()` at the top, since `zip()` will call this anyway.

Reply:
I think its useful to have the functionality of reiterating and iterator, but imo the current version of your code mixes the concepts of `Iterable` and `Iterator` (https://stackoverflow.com/questions/9884132/what-exactly-are-iterator-iterable-and-iteration): the `parallel_iter()` function expects an `Iterable` (something that implements `__iter__` to return an `Iterator`). In your `reiterate=False` case, the function gets passed an `Iterator` (something that implements a `next()` method). Our implementation of `SequenceReader.__iter__()` is a `Generator`, which returns an `Iterator`, so the `iter()` call in `parallel_iter` creates that iterator. I am not sure why we call `iter()` explicitly before the `zip()` but I think there was a reason for it, @tdomhan ?

So in my opinion this new implementation is not properly reflected in the type annotation of `parallel_iter`, and I like the simplicity of the zip-for loop over the while True logic.
How about making `parallel_iter()` a wrapper function that calls `iter()` on the `Iterable` inputs, passing the result to `parallel_iterator()`, containing your current implementation minus the 1st part. Downside is that we loose the nice `zip` & for loop which I like a lot more than this `while True, break` stuff.